src/util/type_traits.h

```diff
@@ -0,0 +1,30 @@
+// Author: Diffblue Ltd.
+
+/// \file
+/// Back ports of utilities available in the `<type_traits>` library for C++14
+/// or C++17 to C++11. These can be replaced with the standard library versions
+/// as and when we upgrade the version CBMC is compiled with.
+
+#ifndef CPROVER_UTIL_TYPE_TRAITS_H
+#define CPROVER_UTIL_TYPE_TRAITS_H
+
+#include <type_traits>
+
+template <bool value>
+using bool_constant = std::integral_constant<bool, value>;
+
+namespace detail // NOLINT
+{
+// Implementation detail of `void_t`.
+template <typename... typest>
+struct make_voidt
+{
+  using type = void;
+};
+} // namespace detail
+
+// Back ported from C++17 STL.
+template <typename... typest>
+using void_t = typename detail::make_voidt<typest...>::type;
+
+#endif // CPROVER_UTIL_TYPE_TRAITS_H
```

Comment:
I'd suggest to move this to the only file actually using it and also `#ifdef` it with compiler versions (`__cplusplus`). Or, at bare minimum, make the comments sufficiently detailed to explain what exactly the code should look like once we move to C++14 (as is done in `file_util.h` or `optional.h`).

Reply:
I am more than happy to add extra comments detailing how to transition to using the C++17 version in preparation for the move to that standard. However given that the back ported version will compile with either standards version, is to worth complicating things with different code for different standards, based on the version in `__cplusplus`?